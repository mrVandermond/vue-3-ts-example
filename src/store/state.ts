import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';

export interface IState {
  favoritePhotos: IPhotoWithIsFavorite[];
  modal: IModalConfig;
  title: ITitleConfig;
}

export const state = (): IState => ({
  favoritePhotos: [],
  modal: {
    isOpen: false,
    url: '',
  },
  title: {
    isShow: false,
    text: '',
    xCoord: 0,
    yCoord: 0,
  },
});
