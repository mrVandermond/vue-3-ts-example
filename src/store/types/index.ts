/* mutations */
export enum EMutations {
  ADD_TO_FAVORITES = 'ADD_TO_FAVORITES',
  REMOVE_FROM_FAVORITES = 'REMOVE_FROM_FAVORITES',
  SET_MODAL_CONFIG = 'SET_MODAL_CONFIG',
  SET_TITLE_CONFIG = 'SET_TITLE_CONFIG',
}
/* mutations */
/* getters */
export enum EGetters {
  IS_FAVORITE_PHOTO_BY_ID = 'IS_FAVORITE_PHOTO_BY_ID',
}
/* getters */
/* actions */
export enum EActions {
  GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE = 'GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE',
  SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE = 'SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE',
  FETCH_USERS = 'FETCH_USERS',
  FETCH_ALBUMS_BY_USER_ID = 'FETCH_ALBUMS_BY_USER_ID',
  FETCH_PHOTOS_BY_ALBUM_ID = 'FETCH_PHOTOS_BY_ALBUM_ID',
}
/* actions */

export const storeTypes = {
  EMutations,
  EGetters,
  EActions,
};
