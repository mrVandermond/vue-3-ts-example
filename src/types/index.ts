interface IUserCompany {
  bs: string;
  catchPhrase: string;
  name: string;
}

interface IUserAddress {
  city: string;
  street: string;
  suite: string;
  zipcode: string;
}

export interface IUser {
  name: string;
  phone: string;
  username: string;
  website: string;
  id: number;
  email: string;
  company: IUserCompany;
  address: IUserAddress;
}

export interface IAlbum {
  id: number;
  title: string;
  userId: number;
}

export interface ITab {
  text: string;
  routeName: string;
  isActive: boolean;
}

export interface IPhoto {
  albumId: number;
  id: number;
  thumbnailUrl: string;
  title: string;
  url: string;
}

export interface IPhotoWithIsFavorite extends IPhoto {
  isFavorite: boolean;
}

export interface IModalConfig {
  isOpen: boolean;
  url: string;
}

export interface ITitleConfig {
  isShow: boolean;
  text: string;
  xCoord: number;
  yCoord: number;
}
