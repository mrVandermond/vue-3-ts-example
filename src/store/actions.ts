import { storeTypes } from '@/store/types';
import { utlCst } from '@/utils';
import type { Mutations } from '@/store/mutations';
import type { ActionContext, ActionTree } from 'vuex';
import type { IState } from '@/store/state';
import type { IAlbum, IPhoto, IPhotoWithIsFavorite, IUser } from '@/types';
import * as api from '@/api';

type AugmentedActionContext = {
  commit<K extends keyof Mutations>(
    key: K,
    payload: Parameters<Mutations[K]>[1]
  ): ReturnType<Mutations[K]>;
} & Omit<ActionContext<IState, IState>, 'commit'>;

export interface Actions {
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.FETCH_USERS](
    ctx: AugmentedActionContext,
  ): Promise<IUser[]>;
  [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IAlbum[]>;
  [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IPhoto[]>;
}

export const actions: ActionTree<IState, IState> & Actions = {
  async [storeTypes.EActions.FETCH_USERS]() {
    const { data } = await api.fetchUsers();

    return data;
  },
  async [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](ctx, userId) {
    const { data } = await api.fetchAlbumsByUserId(userId);

    return data;
  },
  async [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](ctx, albumId) {
    const { data } = await api.fetchPhotosByAlbumId(albumId);

    return data;
  },
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE]({ commit }) {
    const jsonFavoritePhotos = localStorage.getItem(utlCst.LOCAL_STORAGE_KEY);

    if (!jsonFavoritePhotos) return;

    const favoritePhotos = (JSON.parse(jsonFavoritePhotos) as IPhotoWithIsFavorite[]);

    commit(storeTypes.EMutations.ADD_TO_FAVORITES, favoritePhotos);
  },
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE]({ state }) {
    const { favoritePhotos } = state;
    const jsonFavoritePhotos = JSON.stringify(favoritePhotos);

    localStorage.setItem(utlCst.LOCAL_STORAGE_KEY, jsonFavoritePhotos);
  },
};
