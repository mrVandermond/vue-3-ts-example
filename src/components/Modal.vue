<template>
  <div
    v-if="isOpen"
    class="modal-wrapper"
  >
    <div class="modal-overlay" />

    <div
      class="close-icon"
      @click="onClickClose"
    />

    <div class="modal">
      <img
        :src="url"
        alt="modal-photo-image"
      >
    </div>
  </div>
</template>

<script lang="ts" setup>
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';
import { computed, watchEffect } from 'vue';

const store = customUseStore();
const isOpen = computed(() => store.state.modal.isOpen);
const url = computed(() => store.state.modal.url);

watchEffect(() => {
  if (isOpen.value) {
    document.body.style.overflow = 'hidden';
  } else {
    document.body.style.overflow = '';
  }
});

const onClickClose = (): void => {
  store.commit(storeTypes.EMutations.SET_MODAL_CONFIG, {
    url: '',
    isOpen: false,
  });
};
</script>

<style lang="scss">
.modal-wrapper {
  z-index: 1000;
  position: absolute;
  width: 100vw;
  height: 100vh;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
}
.modal-overlay {
  position: absolute;
  top: 0;
  left: 0;
  background-color: $black;
  opacity: 0.47;
  width: 100vw;
  height: 100vh;
}
.modal {
  z-index: 1002;
}
.close-icon {
  position: absolute;
  top: 32px;
  right: 32px;
  width: 24px;
  height: 24px;
  cursor: pointer;

  &:before,
  &:after {
    position: absolute;
    background-color: $white;
    content: '';
    transform: rotateZ(45deg);
  }

  &:before {
    top: 0;
    left: calc(50% - 1px);
    height: 100%;
    width: 2px;
  }

  &:after {
    top: calc(50% - 1px);
    left: 0;
    width: 100%;
    height: 2px;
  }
}
</style>
