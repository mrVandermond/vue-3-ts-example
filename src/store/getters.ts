import type { GetterTree } from 'vuex';
import type { IState } from '@/store/state';
import { storeTypes } from '@/store/types';

export interface Getters {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](state: IState): (photoId: number) => boolean;
}

export const getters: GetterTree<IState, IState> & Getters = {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID]: (state) => (photoId: number) => {
    const photo = state.favoritePhotos.find((item) => item.id === photoId);

    return !!photo;
  },
};
