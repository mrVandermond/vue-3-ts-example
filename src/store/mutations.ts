import { storeTypes } from '@/store/types';
import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';
import type { IState } from '@/store/state';
import type { MutationTree } from 'vuex';
import { utlFuncs } from '@/utils';

export interface Mutations<S = IState> {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state: S, payload: IPhotoWithIsFavorite[]): void;
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state: S, payload: IPhotoWithIsFavorite): void;
  [storeTypes.EMutations.SET_MODAL_CONFIG](state: S, payload: IModalConfig): void;
  [storeTypes.EMutations.SET_TITLE_CONFIG](state: S, payload: ITitleConfig): void;
}

export const mutations: MutationTree<IState> & Mutations = {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state, photos: IPhotoWithIsFavorite[]) {
    const photosWithoutExisting = photos.filter((photo) => (
      utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id') === -1
    ));

    state.favoritePhotos.push(...photosWithoutExisting.map((photo) => ({ ...photo, isFavorite: true })));
  },
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state, photo: IPhotoWithIsFavorite) {
    const photoIndex = utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id');

    if (photoIndex === -1) return;

    state.favoritePhotos.splice(photoIndex, 1);
  },
  [storeTypes.EMutations.SET_MODAL_CONFIG](state, config) {
    state.modal.isOpen = config.isOpen;
    state.modal.url = config.url;
  },
  [storeTypes.EMutations.SET_TITLE_CONFIG](state, config) {
    state.title.isShow = config.isShow;
    state.title.text = config.text;
    state.title.xCoord = config.xCoord;
    state.title.yCoord = config.yCoord;
  },
};
