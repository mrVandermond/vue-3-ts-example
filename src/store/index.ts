import type { StoreOptions } from 'vuex';
import { createStore, useStore as baseUseStore } from 'vuex';
import type { InjectionKey } from 'vue';
import type { Store } from '@/store/types/overrides';
import type { IState } from './state';
import { state } from './state';
import { mutations } from './mutations';
import { getters } from './getters';
import { actions } from './actions';

export const key: InjectionKey<Store> = Symbol('store');

export function customUseStore(): Store {
  return baseUseStore(key);
}

export function customCreateStore(options: StoreOptions<IState>): Store {
  return createStore<IState>(options);
}

export const store = customCreateStore({
  state,
  mutations,
  getters,
  actions,
});
