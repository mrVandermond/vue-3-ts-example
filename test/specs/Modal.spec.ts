import { nextTick } from 'vue';

import Modal from '@/components/Modal.vue';

import type { Store } from '@/store/types/overrides';

import { customCreateStore, key } from '@/store';
import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { getters } from '@/store/getters';
import { actions } from '@/store/actions';
import { storeTypes } from '@/store/types';

import type { TWrapperFactoryReturnType } from '../utils';
import { wrapperFactory } from '../utils';

describe('Modal component', () => {
  let store: Store;

  beforeEach(() => {
    store = customCreateStore({
      state,
      mutations,
      getters,
      actions,
    });

    jest.spyOn(store, 'commit');
  });

  const getWrapper = (): TWrapperFactoryReturnType => wrapperFactory(Modal, {
    global: {
      plugins: [[store, key]],
    },
  });

  it('Открытая модалка', async () => {
    const wrapper = getWrapper();

    store.commit(storeTypes.EMutations.SET_MODAL_CONFIG, {
      isOpen: true,
      url: 'https://via.placeholder.com/600/771796',
    });

    await nextTick();

    expect(document.body.style.overflow).toEqual('hidden');

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Закрытие модалки', async () => {
    const wrapper = getWrapper();

    store.commit(storeTypes.EMutations.SET_MODAL_CONFIG, {
      isOpen: true,
      url: 'https://via.placeholder.com/600/771796',
    });

    await nextTick();

    const closeButton = wrapper.find('.close-icon');

    await closeButton.trigger('click');

    expect(document.body.style.overflow).toEqual('');
    expect(store.commit).toHaveBeenCalledWith(storeTypes.EMutations.SET_MODAL_CONFIG, {
      isOpen: false,
      url: '',
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
