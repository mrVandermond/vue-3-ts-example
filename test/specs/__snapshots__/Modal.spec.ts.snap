// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal component > Закрытие модалки 1`] = `"<!--v-if-->"`;

exports[`Modal component > Открытая модалка 1`] = `
"<div class=\\"modal-wrapper\\">
  <div class=\\"modal-overlay\\"></div>
  <div class=\\"close-icon\\"></div>
  <div class=\\"modal\\"><img src=\\"https://via.placeholder.com/600/771796\\" alt=\\"modal-photo-image\\"></div>
</div>"
`;
